File: src/types/conversation.ts
export type ConversationThreadStatus =
    | 'idle'
    | 'running'
    | 'interrupted'
    | 'completed'
    | 'failed';

export type ConversationTurnStatus =
    | 'queued'
    | 'running'
    | 'waiting_approval'
    | 'interrupted'
    | 'completed'
    | 'failed';

export type ConversationRole = 'user' | 'assistant' | 'system' | 'tool';

export type ConversationItemType =
    | 'message'
    | 'reasoning'
    | 'command'
    | 'file_change'
    | 'mcp_tool'
    | 'approval'
    | 'error'
    | 'raw';

export type ConversationItemStatus = 'pending' | 'running' | 'completed' | 'failed';

export type ConversationEventType =
    | 'thread_started'
    | 'thread_resumed'
    | 'turn_started'
    | 'turn_interrupted'
    | 'turn_completed'
    | 'turn_failed'
    | 'item_started'
    | 'item_delta'
    | 'item_completed'
    | 'approval_requested'
    | 'approval_resolved';

export type ApprovalRequestType = 'command' | 'file_change' | 'mcp_tool' | 'user_input';
export type ApprovalDecision = 'approved' | 'denied';
export type CodexMcpServerState = 'unknown' | 'connected' | 'disconnected' | 'error';

export interface ThreadStartInput {
    workspaceId?: string;
    cwd: string;
    model?: string;
    providerId?: string;
    approvalPolicy?: string;
    sandboxPolicy?: string;
    metadata?: Record<string, unknown>;
}

export interface ThreadResumeInput {
    threadId: string;
    workspaceId?: string;
    cwd?: string;
    sourcePath?: string;
    metadata?: Record<string, unknown>;
}

export interface TurnStartInput {
    threadId: string;
    prompt: string;
    model?: string;
    approvalPolicy?: string;
    sandboxPolicy?: string;
    metadata?: Record<string, unknown>;
}

export interface ApprovalResponseInput {
    approvalId: string;
    decision: ApprovalDecision;
    message?: string;
    metadata?: Record<string, unknown>;
}

export interface ConversationThread {
    id: string;
    workspaceId?: string;
    cwd: string;
    title?: string;
    status: ConversationThreadStatus;
    createdAt: number;
    updatedAt: number;
}

export interface ConversationTurn {
    id: string;
    threadId: string;
    status: ConversationTurnStatus;
    createdAt: number;
    completedAt?: number;
}

export interface ConversationItem {
    id: string;
    threadId: string;
    turnId?: string;
    itemType: ConversationItemType;
    role?: ConversationRole;
    status: ConversationItemStatus;
    content?: string;
    summary?: string;
    metadata: Record<string, unknown>;
    createdAt: number;
    completedAt?: number;
}

export interface ApprovalRequest {
    id: string;
    threadId: string;
    turnId?: string;
    itemId?: string;
    requestType: ApprovalRequestType;
    title: string;
    body?: string;
    command?: string;
    cwd?: string;
    toolName?: string;
    metadata: Record<string, unknown>;
    createdAt: number;
}

export interface ConversationEvent {
    id: string;
    threadId: string;
    turnId?: string;
    approvalId?: string;
    eventType: ConversationEventType;
    item?: ConversationItem;
    approvalRequest?: ApprovalRequest;
    delta?: string;
    metadata: Record<string, unknown>;
    createdAt: number;
}

export interface ConversationThreadSnapshot {
    thread: ConversationThread;
    items: ConversationItem[];
    pendingApprovals: ApprovalRequest[];
}

export interface CodexConfigSummary {
    workspaceId?: string;
    codexHome?: string;
    configPath?: string;
    configExists: boolean;
    model?: string;
    providerId?: string;
    approvalPolicy?: string;
    sandboxPolicy?: string;
    metadata: Record<string, unknown>;
}

export interface CodexModelInfo {
    id: string;
    name: string;
    providerId?: string;
    supportsReasoning: boolean;
    supportsTools: boolean;
}

export interface CodexMcpServerStatus {
    serverId: string;
    name: string;
    state: CodexMcpServerState;
    message?: string;
}

File: src-tauri/src/models/conversation.rs
#![allow(dead_code)]

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

pub type JsonObject = Map<String, Value>;

fn default_json_object() -> JsonObject {
    JsonObject::new()
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ConversationThreadStatus {
    Idle,
    Running,
    Interrupted,
    Completed,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ConversationTurnStatus {
    Queued,
    Running,
    WaitingApproval,
    Interrupted,
    Completed,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ConversationRole {
    User,
    Assistant,
    System,
    Tool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ConversationItemType {
    Message,
    Reasoning,
    Command,
    FileChange,
    McpTool,
    Approval,
    Error,
    Raw,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ConversationItemStatus {
    Pending,
    Running,
    Completed,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ConversationEventType {
    ThreadStarted,
    ThreadResumed,
    TurnStarted,
    TurnInterrupted,
    TurnCompleted,
    TurnFailed,
    ItemStarted,
    ItemDelta,
    ItemCompleted,
    ApprovalRequested,
    ApprovalResolved,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ApprovalRequestType {
    Command,
    FileChange,
    McpTool,
    UserInput,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ApprovalDecision {
    Approved,
    Denied,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ThreadStartInput {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub workspace_id: Option<String>,
    pub cwd: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approval_policy: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sandbox_policy: Option<String>,
    #[serde(default = "default_json_object")]
    pub metadata: JsonObject,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ThreadResumeInput {
    pub thread_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub workspace_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cwd: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_path: Option<String>,
    #[serde(default = "default_json_object")]
    pub metadata: JsonObject,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct TurnStartInput {
    pub thread_id: String,
    pub prompt: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approval_policy: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sandbox_policy: Option<String>,
    #[serde(default = "default_json_object")]
    pub metadata: JsonObject,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ApprovalResponseInput {
    pub approval_id: String,
    pub decision: ApprovalDecision,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    #[serde(default = "default_json_object")]
    pub metadata: JsonObject,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ConversationThread {
    pub id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub workspace_id: Option<String>,
    pub cwd: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    pub status: ConversationThreadStatus,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ConversationTurn {
    pub id: String,
    pub thread_id: String,
    pub status: ConversationTurnStatus,
    pub created_at: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ConversationItem {
    pub id: String,
    pub thread_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub turn_id: Option<String>,
    pub item_type: ConversationItemType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub role: Option<ConversationRole>,
    pub status: ConversationItemStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
    #[serde(default = "default_json_object")]
    pub metadata: JsonObject,
    pub created_at: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ApprovalRequest {
    pub id: String,
    pub thread_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub turn_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub item_id: Option<String>,
    pub request_type: ApprovalRequestType,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub body: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub command: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cwd: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_name: Option<String>,
    #[serde(default = "default_json_object")]
    pub metadata: JsonObject,
    pub created_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ConversationEvent {
    pub id: String,
    pub thread_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub turn_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approval_id: Option<String>,
    pub event_type: ConversationEventType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub item: Option<ConversationItem>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approval_request: Option<ApprovalRequest>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub delta: Option<String>,
    #[serde(default = "default_json_object")]
    pub metadata: JsonObject,
    pub created_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ConversationThreadSnapshot {
    pub thread: ConversationThread,
    pub items: Vec<ConversationItem>,
    pub pending_approvals: Vec<ApprovalRequest>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CodexConfigSummary {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub workspace_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub codex_home: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub config_path: Option<String>,
    pub config_exists: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approval_policy: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sandbox_policy: Option<String>,
    #[serde(default = "default_json_object")]
    pub metadata: JsonObject,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CodexModelInfo {
    pub id: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider_id: Option<String>,
    pub supports_reasoning: bool,
    pub supports_tools: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CodexMcpServerState {
    Unknown,
    Connected,
    Disconnected,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CodexMcpServerStatus {
    pub server_id: String,
    pub name: String,
    pub state: CodexMcpServerState,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::{ThreadStartInput, TurnStartInput};
    use serde_json::json;

    #[test]
    fn thread_start_input_uses_camel_case_fields() {
        let input = ThreadStartInput {
            workspace_id: Some("workspace-1".to_string()),
            cwd: "C:\\guodevelop\\ccg-switch".to_string(),
            model: Some("gpt-5".to_string()),
            provider_id: Some("provider-1".to_string()),
            approval_policy: Some("on-request".to_string()),
            sandbox_policy: Some("workspace-write".to_string()),
            metadata: serde_json::Map::new(),
        };

        let value = serde_json::to_value(input).expect("thread input should serialize");
        assert_eq!(value["workspaceId"], "workspace-1");
        assert_eq!(value["providerId"], "provider-1");
        assert!(value.get("workspace_id").is_none());
    }

    #[test]
    fn turn_start_input_rejects_unknown_fields() {
        let result = serde_json::from_value::<TurnStartInput>(json!({
            "threadId": "thread-1",
            "prompt": "hello",
            "unknownField": true
        }));

        assert!(result.is_err());
    }
}
